sdk/core/Azure.Core/src/Pipeline/HttpPipeline.cs

```diff
@@ -124,7 +129,68 @@ public Response SendRequest(Request request, CancellationToken cancellationToken
         /// </example>
         public static IDisposable CreateClientRequestIdScope(string? clientRequestId)
         {
-            return ReadClientRequestIdPolicy.StartScope(clientRequestId);
+            return CreateHttpMessagePropertiesScope(new Dictionary<string, object?>() { { ReadClientRequestIdPolicy.MessagePropertyKey, clientRequestId } });
+        }
+
+        /// <summary>
+        /// Creates a scope in which all <see cref="HttpMessage"/>s would have provided properties.
+        /// </summary>
+        /// <param name="messageProperties">Properties to be added to <see cref="HttpMessage"/>s</param>
+        /// <returns>The <see cref="IDisposable"/> instance that needs to be disposed when properties shouldn't be used anymore.</returns>
+        public static IDisposable CreateHttpMessagePropertiesScope(IDictionary<string, object?> messageProperties)
+        {
+            Argument.AssertNotNull(messageProperties, nameof(messageProperties));
+            CurrentHttpMessagePropertiesScope.Value = new HttpMessagePropertiesScope(messageProperties, CurrentHttpMessagePropertiesScope.Value);
+            return CurrentHttpMessagePropertiesScope.Value;
+        }
+
+        private static void AddHttpMessageProperties(HttpMessage message)
+        {
+            if (CurrentHttpMessagePropertiesScope.Value != null)
+            {
+                foreach (KeyValuePair<string, object?> kvp in CurrentHttpMessagePropertiesScope.Value.Properties)
+                {
+                    if (kvp.Value != null)
+                    {
+                        message.SetProperty(kvp.Key, kvp.Value);
+                    }
+                }
+            }
+        }
+
+        private class HttpMessagePropertiesScope : IDisposable
+        {
+            private readonly HttpMessagePropertiesScope? _parent;
+            private bool _disposed;
+
+            internal HttpMessagePropertiesScope(IDictionary<string, object?> messageProperties, HttpMessagePropertiesScope? parent)
+            {
+                if (parent != null)
+                {
+                    Properties = new Dictionary<string, object?>(parent.Properties);
+                    foreach (var kvp in messageProperties)
+                    {
+                        Properties[kvp.Key] = kvp.Value;
+                    }
+                }
+                else
+                {
+                    Properties = new Dictionary<string, object?>(messageProperties);
+                }
+                _parent = parent;
+            }
+
+            public Dictionary<string, object?> Properties { get; }
+
+            public void Dispose()
+            {
+                if (_disposed)
+                {
+                    return;
+                }
+                CurrentHttpMessagePropertiesScope.Value = _parent;
+                _disposed = true;
```

Comment:
Do we need to worry about being more thread-safe here?  Should we use something in Interlocked to set _disposed = true?

Reply:
there doesn't seem to be a valid scenario where someone would want to dispose it from multiple threads.